sql: improve performance of userCanSeeDescriptor

rafiss opened the pull request:
### rttanalysis: run benchmarks with non-root user

The root user skips certain privilege checks, so using it in these
benchmarks would create results that don't reflect the normal experience
that most users have. The benchmark expectations are rewritten
accordingly.

At the same time, this adds new tests for queries run by npgsql that
were found to be slow.

---

### sql: improve performance of userCanSeeDescriptor

28bcaaf (and the related backports)
changed the order that checks were made when determining if a user can
view a descriptor for the purpose of populating a virtual table. It
improved the performance with virtual table descriptors, but made things
worse for other descriptors.

The performance had degraded enough to the point that the nightly npgsql
tests began failing.

This commit goes back to the previous order of checking privileges, and
also adds another check so that virtual table descriptors can still
short circuit.

```
before
BenchmarkORMQueries/npgsql_types-10         	      14	  72771417 ns/op	        60.57 roundtrips
BenchmarkORMQueries/npgsql_fields-10        	       4	 288099771 ns/op	        84.00 roundtrips

after
BenchmarkORMQueries/npgsql_types-10         	      30	  37119303 ns/op	        13.27 roundtrips
BenchmarkORMQueries/npgsql_fields-10        	      72	  16410549 ns/op	         9.056 roundtrips
```

---

fixes https://github.com/cockroachdb/cockroach/issues/109960
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111819)
<!-- Reviewable:end -->

rafiss committed:
rttanalysis: run benchmarks with non-root user

The root user skips certain privilege checks, so using it in these
benchmarks would create results that don't reflect the normal experience
that most users have. The benchmark expectations are rewritten
accordingly.

At the same time, this adds new tests for queries run by npgsql that
were found to be slow.

Release note: None

rafiss committed:
sql: improve performance of userCanSeeDescriptor

28bcaaf (and the related backports)
changed the order that checks were made when determining if a user can
view a descriptor for the purpose of populating a virtual table. It
improved the performance with virtual table descriptors, but made things
worse for other descriptors.

The performance had degraded enough to the point that the nightly npgsql
tests began failing.

This commit goes back to the previous order of checking privileges, and
also adds another check so that virtual table descriptors can still
short circuit.

before
BenchmarkORMQueries/npgsql_types-10         	      14	  72771417 ns/op	        60.57 roundtrips
BenchmarkORMQueries/npgsql_fields-10        	       4	 288099771 ns/op	        84.00 roundtrips

after
BenchmarkORMQueries/npgsql_types-10         	      30	  37119303 ns/op	        13.27 roundtrips
BenchmarkORMQueries/npgsql_fields-10        	      72	  16410549 ns/op	         9.056 roundtrips

Release note: None

fqazi reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>  Nice find!

> Reviewed 8 of 8 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111819#-:-Ng08dww-x9CIV7W5RkS:b4csktu)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss commented:
tftr!

bors r+
